source/module_hsolver/diago_cusolver.cpp

```diff
@@ -135,45 +135,59 @@ namespace hsolver
         ModuleBase::timer::tick("DiagoCusolver", "cusolver");
 
 #ifdef __MPI
-        // gather matrices from processes to root process
-        gatherMatrix(myid, root_proc, h_mat, h_mat_g);
-        gatherMatrix(myid, root_proc, s_mat, s_mat_g);
+        if(num_procs > 1)
+        {
+            // gather matrices from processes to root process
+            gatherMatrix(myid, root_proc, h_mat, h_mat_g);
+            gatherMatrix(myid, root_proc, s_mat, s_mat_g);
+        }
 #endif
 
         // Call the dense diagonalization routine
 #ifdef __MPI
-        MPI_Barrier(MPI_COMM_WORLD);
-        if (myid == root_proc)
+        if(num_procs > 1)
         {
-            psi_g = new T[h_mat_g.row * h_mat_g.col];
-            this->dc.Dngvd(h_mat_g.col, h_mat_g.row, h_mat_g.p, s_mat_g.p, eigen.data(), psi_g);
+            MPI_Barrier(MPI_COMM_WORLD);
+            if (myid == root_proc)
+            {
+                psi_g = new T[h_mat_g.row * h_mat_g.col];
+                this->dc.Dngvd(h_mat_g.col, h_mat_g.row, h_mat_g.p, s_mat_g.p, eigen.data(), psi_g);
+            }
+            else
+            {
+                psi_g = new T[1];
+            }
+            MPI_Barrier(MPI_COMM_WORLD);
+            
+            // broadcast eigenvalues to all processes
+            MPI_Bcast(eigen.data(), GlobalV::NBANDS, MPI_DOUBLE, root_proc, MPI_COMM_WORLD);
+
+            // distribute psi to all processes
+            distributePsi(this->ParaV->desc_wfc, psi.get_pointer(), psi_g);
         }
         else
         {
-            psi_g = new T[1];
+            this->dc.Dngvd(h_mat.row, h_mat.col, h_mat.p, s_mat.p, eigen.data(), psi.get_pointer());
         }
-        MPI_Barrier(MPI_COMM_WORLD);
-        // broadcast eigenvalues to all processes
-        MPI_Bcast(eigen.data(), GlobalV::NBANDS, MPI_DOUBLE, root_proc, MPI_COMM_WORLD);
-
-        // distribute psi to all processes
-        distributePsi(this->ParaV->desc_wfc, psi.get_pointer(), psi_g);
 #else
         // Call the dense diagonalization routine
         this->dc.Dngvd(h_mat.row, h_mat.col, h_mat.p, s_mat.p, eigen.data(), psi.get_pointer());
 #endif
         // Stop the timer for the cusolver operation
         ModuleBase::timer::tick("DiagoCusolver", "cusolver");
 
-        // Copy the eigenvalues and eigenvectors to the output arrays
+        // Copy the eigenvalues to the output arrays
         const int inc = 1;
         BlasConnector::copy(GlobalV::NBANDS, eigen.data(), inc, eigenvalue_in, inc);
 
         // Free allocated memory
 #ifdef __MPI
-        delete[] h_mat_g.p;
-        delete[] s_mat_g.p;
-        delete[] psi_g;
+        if(num_procs > 1)
+        {
+            delete[] h_mat_g.p;
```

Comment:
you can try: use stl vector and avoid new and delete

Reply:
The type of `h_mat_g` is MatrixBlock, which is a structure widely used throughout the program, so change the type of h_mat_g.p would incur significant costs.
![image](https://github.com/deepmodeling/abacus-develop/assets/153698752/2168d78e-3668-4757-9431-fcb2a4bdef1b)

Reply:
try harder to resolve the issue